File: backend/src/main/kotlin/ai/project/rio/cardtransaction/CardTransaction.kt
package ai.project.rio.cardtransaction

import ai.project.rio.money.Money
import java.time.Instant

enum class CardTransactionType { CREDIT, DEBIT }

enum class CardTransactionStatus { PENDING, COMPLETED, DECLINED }

/**
 * `amount` is a magnitude and is always > 0 for persisted card transactions.
 * Direction (money in vs. money out) comes from `type`, never from the sign of the amount.
 */
data class CardTransaction(
    val id: String,
    val description: String,
    val amount: Money,
    val type: CardTransactionType,
    val status: CardTransactionStatus,
    val createdAt: Instant,
)

/**
 * Whether a create request was one object or an array of them. The two answer with different
 * response shapes, so the shape is part of what an Idempotency-Key identifies.
 */
enum class RequestShape { ONE, MANY }

/** What a caller supplies to create a card transaction; the service assigns id, status, and createdAt. */
data class NewCardTransaction(
    val description: String,
    val amount: Money,
    val type: CardTransactionType,
)

File: frontend/src/types/cardTransaction.ts
import type { Money } from "../money/money";

// Domain types used by pages and components. Money is already converted to bigint here.

export type CardTransactionType = "CREDIT" | "DEBIT";

export type CardTransactionStatus = "PENDING" | "COMPLETED" | "DECLINED";

export interface CardTransaction {
  id: string;
  description: string;
  /** Magnitude, always > 0. Direction comes from `type`. */
  amount: Money;
  type: CardTransactionType;
  status: CardTransactionStatus;
  createdAt: Date;
  /** The validated API instant, verbatim (e.g. `2026-09-02T15:30:00.123456789Z`). Date truncates to milliseconds; use this for exact display and log correlation. */
  createdAtInstant: string;
}

export interface CreateCardTransactionInput {
  description: string;
  amount: Money;
  type: CardTransactionType;
}
